libs/common/src/test/java20/org/opensearch/common/round/BtreeSearcherTests.java

```diff
@@ -0,0 +1,20 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.round;
+
+import java.util.List;
+
+import jdk.incubator.vector.LongVector;
+
+public class BtreeSearcherTests extends RoundableTestCase {
+    @Override
+    public Roundable newInstance(long[] values, int size) {
+        return new BtreeSearcher(values, size, randomFrom(List.of(LongVector.SPECIES_128, LongVector.SPECIES_256, LongVector.SPECIES_512)));
```

Comment:
I believe the test case does not really reflect what the code in production will do: the species is not configurable right now and always use `LongVector.SPECIES_PREFERRED`, I think we should either drop the species parametrization from test case (that would simplify build a lot) or allow it to configurable (with the system property).

Reply:
My intent is to have unit-tests cover vector sizes for different platforms, since the number of vector lanes determines the block size, which affects the memory layout and calculation of offsets. Without parameterizing this, the unit-tests will only test with the vector size supported by the build machine.

It's true that in production, `LongVector.SPECIES_PREFERRED` should be the only valid option. I'm enforcing this with package-protected visibility of the constructor and relying on RoundableFactory to create the _best_ instance.

As a side note, even if the build machine supports up to `LongVector.SPECIES_256`, we can still execute our code perfectly fine with `LongVector.SPECIES_512`. Java takes care of making it portable, though, it won't be the fastest on that hardware.

Reply:
> My intent is to have unit-tests cover vector sizes for different platforms, since the number of vector lanes determines the block size, which affects the memory layout and calculation of offsets.

I do see pros and cons here. The first point is that not every platform support all species and we may very likely get bug reports from people that tests fail on their specific hardware. The second point is that we never let people choose. Testing the production like flow (with `LongVector.SPECIES_PREFERRED`) would remove the need to branch test cases (and complicate the build).

Reply:
> The first point is that not every platform support all species and we may very likely get bug reports from people that tests fail on their specific hardware.

Tests won't fail. If we use a vector species not _natively_ supported by some hardware, everything works fine but slower. Java takes care of making the code portable.

> The second point is that we never let people choose.

I'm with you on this. This isn't user-customizable; having this parameter is only to assist with writing the unit tests.

Reply:
> I'm with you on this. This isn't user-customizable; having this parameter is only to assist with writing the unit tests.

I would suggest to get rid of it (and remove `src/test/java20` altogether), let's do test what we would be running

Reply:
May be I'm missing something here. Removing `src/test/java20` would mean no test coverage for `BtreeSearcher`. Is that what you're suggesting?

Reply:
> May be I'm missing something here. Removing src/test/java20 would mean no test coverage for BtreeSearcher. Is that what you're suggesting?

Not at all, the test would pick up the source from multi release, so if we run the test under JDK-20 with SIMD feature enabled - we would have to use the BtreeSearcher

Reply:
Okay, I think I got what you're suggesting. Instead of having separate tests for each source, use the code path executed in production, and let the build matrix execute the tests with different JDK versions.

However, there are a few more challenges:
1. The unit tests in `src/test/java` don't execute with the MR JAR. It'll need changes similar to [this](https://github.com/opensearch-project/OpenSearch/pull/11194/files#diff-332bcff66712f49326d714f043d507a7d3a203bd829d00a84be17785f1045342R74-R75) (this is doable).
2. Since `src/test/java` compiles for Java 11, it won't be possible to use the Vector API directly. Either we write code that doesn't refer to the higher-level language features or we rely on extensive use of reflection similar to [this](https://github.com/opensearch-project/OpenSearch/pull/11194/files#diff-7d0a120f2782ea93809fd4ece3883a9ee3a6e5ef3af315d25e772ac1fa50644eR143).
3. Expanding on (2), `src/test/java` should be the tests for the main implementation in `src/main/java`. It shouldn't know anything outside of this.

Reply:
Thanks @ketanv3 

> 1. The unit tests in src/test/java don't execute with the MR JAR. It'll need changes similar to [this](https://github.com/opensearch-project/OpenSearch/pull/11194/files#diff-332bcff66712f49326d714f043d507a7d3a203bd829d00a84be17785f1045342R74-R75) (this is doable).

:+1: 

> 2. Since src/test/java compiles for Java 11, it won't be possible to use the Vector API directly. Either we write code that doesn't refer to the higher-level language features or we rely on extensive use of reflection similar to [this](https://github.com/opensearch-project/OpenSearch/pull/11194/files#diff-7d0a120f2782ea93809fd4ece3883a9ee3a6e5ef3af315d25e772ac1fa50644eR143).

No, the API level does not matter here, the runtime JDK does. The way it should work is this:
 - if runtime JDK < 20 - nothing happens
 - if runtime JDK > 20 - the optimization is consider (taking system property into account)

I could certainly help you here. Since we build with JDK-21 on Jenkins, we should be all set to have multi-release JAR available.

> 3. Expanding on (2), src/test/java should be the tests for the main implementation in src/main/java. It shouldn't know anything outside of this.

Correct, we run test under JDK-21 now (on `main`) but we could expand it to JDK-17, I could help you with that.

Reply:
> No, the API level does not matter here, the runtime JDK does.

Sorry, I didn't distinguish between build and runtime JDK versions. What I meant to say is, even when using JDK 20 at build time, the existing test sources still target `JavaVersion.VERSION_11` ([configured here](https://github.com/opensearch-project/OpenSearch/blob/0b2c07a60252292a5a2e41e3af17412ed940886d/buildSrc/build.gradle#L166-L171)). That means, I won't be able to **compile** the unit-test code using the Vector API without extensive use of reflection.

Attempting to use `--add-modules=jdk.incubator.vector` with Java 11 as the target version fails with this error:
```
error: module not found: jdk.incubator.vector
```

In other words, the optimizations (vector module, Java 20 sources, etc.) are present at runtime, they aren't available at compile time. So for that reason, we need separate sources for unit-tests that target Java 20.

Reply:
I took some inspiration from [mrjar-gradle-plugin](https://github.com/melix/mrjar-gradle-plugin) on organization of sources and their tests.
https://melix.github.io/mrjar-gradle-plugin/0.1/index.html#_organization_of_sources

Reply:
@ketanv3 I think when we address https://github.com/opensearch-project/OpenSearch/pull/11194/files#r1399581574 (which basically means the test would not need to access anything from the JDK incubator module), we could think of simplifying the build as well.